src/client/texturesource.cpp

```diff
@@ -1848,58 +1856,105 @@ bool TextureSource::generateImagePart(std::string_view part_of_name,
 
 #undef CHECK_DIM
 
-/*
-	Calculate the color of a single pixel drawn on top of another pixel.
 
-	This is a little more complicated than just video::SColor::getInterpolated
-	because getInterpolated does not handle alpha correctly.  For example, a
-	pixel with alpha=64 drawn atop a pixel with alpha=128 should yield a
-	pixel with alpha=160, while getInterpolated would yield alpha=96.
-*/
-static inline video::SColor blitPixel(const video::SColor src_c, const video::SColor dst_c, u32 ratio)
+namespace {
+
+/// Draw a source color on top of a destination color
+template <bool overlay>
+void blit_pixel(video::SColor src_col, video::SColor &dst_col)
 {
-	if (dst_c.getAlpha() == 0)
-		return src_c;
-	video::SColor out_c = src_c.getInterpolated(dst_c, (float)ratio / 255.0f);
-	out_c.setAlpha(dst_c.getAlpha() + (255 - dst_c.getAlpha()) *
-		src_c.getAlpha() * ratio / (255 * 255));
-	return out_c;
+	u8 dst_a{static_cast<u8>(dst_col.getAlpha())};
+	if constexpr (overlay) {
+		if (dst_a != 255)
+			// The bottom pixel has transparency -> do nothing
+			return;
+	}
+	u8 src_a{static_cast<u8>(src_col.getAlpha())};
+	if (src_a == 0) {
+		// A fully transparent pixel is on top -> do nothing
+		return;
+	}
+	if (src_a == 255 || dst_a == 0) {
+		// The top pixel is fully opaque or the bottom pixel is
+		// fully transparent -> replace the color
+		dst_col = src_col;
+		return;
+	}
+	struct Color { u8 r, g, b; };
+	Color src{
+		static_cast<u8>(src_col.getRed()),
+		static_cast<u8>(src_col.getGreen()),
+		static_cast<u8>(src_col.getBlue())
+	};
+	Color dst{
+		static_cast<u8>(dst_col.getRed()),
+		static_cast<u8>(dst_col.getGreen()),
+		static_cast<u8>(dst_col.getBlue())
+	};
+	if (dst_a == 255) {
+		// A semi-transparent pixel is on top and an opaque one in
+		// the bottom -> lerp r, g, and b
+		dst.r = (dst.r * (255 - src_a) + src.r * src_a) / 255;
+		dst.g = (dst.g * (255 - src_a) + src.g * src_a) / 255;
+		dst.b = (dst.b * (255 - src_a) + src.b * src_a) / 255;
```

Comment:
No. In C and C++, [small integer operands are promoted to int before the arithmetic operations are executed](https://stackoverflow.com/a/46073296). It would only overflow [if I add an explicit cast to u8 here](https://godbolt.org/z/hqz9zGov3). With explicit casts to int, [the same code as without casts is generated](https://godbolt.org/z/vPEon8cva).

With an explicit cast to u16 before the division (`dr = (u16)(dr * (255 - sa) + sr * sa) / 255;`), the compiler generates different code.
Perhaps the compiler detects that `dr * (255 - sa)` and `sr * sa` both lie in $[0, 65025]$, i.e. fit into an u16, but does not detect that their sum is in $[0, 65025]$, too.

Reply:
Uh oh, I think I had already read about this at least once. Nonetheless, I totally forgot about it. Thanks!

(Additional link: <https://en.cppreference.com/w/cpp/language/operator_arithmetic#Conversions>)